File: 04-batch-performance/ai/java/src/main/java/com/treinamento/batch/Transaction.java
package com.treinamento.batch;

/**
 * Immutable data model representing a single financial transaction.
 *
 * CSV columns (in order):
 *   id, amount, source_account, dest_account, timestamp, category, priority
 */
public record Transaction(
    int id,
    double amount,
    String sourceAccount,
    String destAccount,
    long timestamp,       // epoch millis
    String category,
    int priority          // 1 (lowest) .. 5 (highest)
) {

    /**
     * Parses a CSV line produced by DataGenerator.
     * Expected format: id,amount,source_account,dest_account,timestamp,category,priority
     */
    public static Transaction fromCsv(String line) {
        String[] parts = line.split(",", -1);
        if (parts.length != 7) {
            throw new IllegalArgumentException(
                "Invalid CSV line (expected 7 columns): " + line);
        }
        return new Transaction(
            Integer.parseInt(parts[0].trim()),
            Double.parseDouble(parts[1].trim()),
            parts[2].trim(),
            parts[3].trim(),
            Long.parseLong(parts[4].trim()),
            parts[5].trim(),
            Integer.parseInt(parts[6].trim())
        );
    }

    /** CSV representation (mirrors fromCsv). */
    public String toCsv() {
        return String.join(",",
            String.valueOf(id), String.valueOf(amount),
            sourceAccount, destAccount,
            String.valueOf(timestamp), category, String.valueOf(priority));
    }

    /**
     * Two transactions are considered potential duplicates when they share the
     * same source account, destination account, and amount. The time-window
     * check (within 5 seconds) is done in BatchProcessor.
     */
    public boolean isSameOperation(Transaction other) {
        return this.sourceAccount.equals(other.sourceAccount)
            && this.destAccount.equals(other.destAccount)
            && Double.compare(this.amount, other.amount) == 0;
    }
}

File: 04-batch-performance/manual/kotlin/src/main/kotlin/com/treinamento/batch/Transaction.kt
package com.treinamento.batch

data class Transaction(
    val id: Int,
    val amount: Double,
    val sourceAccount: String,
    val destAccount: String,
    val timestamp: Long,
    val category: String,
    val priority: Int
) {
    companion object {
        fun fromCsvLine(line: String): Transaction {
            val parts = line.split(",")
            require(parts.size == 7) { "Expected 7 CSV columns but got ${parts.size}: '$line'" }
            return Transaction(
                id = parts[0].trim().toInt(),
                amount = parts[1].trim().toDouble(),
                sourceAccount = parts[2].trim(),
                destAccount = parts[3].trim(),
                timestamp = parts[4].trim().toLong(),
                category = parts[5].trim(),
                priority = parts[6].trim().toInt()
            )
        }
    }

    fun toCsvLine(): String = "$id,$amount,$sourceAccount,$destAccount,$timestamp,$category,$priority"
}
